Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp

```diff
@@ -361,6 +364,20 @@ static void addParametersShared(const LocalFrame* frame, NetworkResourceLoadPara
 
     if (isMainFrameNavigation)
         parameters.linkPreconnectEarlyHintsEnabled = mainFrame->settings().linkPreconnectEarlyHintsEnabled();
+    else if (RefPtr localMainFrame = dynamicDowncast<LocalFrame>(mainFrame)) {
+        if (auto* document = frame->document(); localMainFrame->document()
+            && isRequestCrossOrigin(&localMainFrame->document()->securityOrigin(), parameters.request.url(), { })
+            && document) {
+
+            std::optional<SecurityOriginData> securityOrigin;
+            if (cachedResource && cachedResource->type() == CachedResource::Type::MainResource)
+                securityOrigin = SecurityOriginData::fromURL(parameters.request.url());
```

Comment:
Permission policy is usually checking the origin of the context doing the work be it capture or networking. I do not see why privacy tokens are special here and why the request's origin is used here.

Reply:
The behavior was surprising to me without this change, but after thinking about this more I agree that the current behavior is correct. Permission was denied when we requested the main resource for a new cross-origin iframe without specifying the parent frame's origin in the allowlist. This is because the `document`'s origin at the time that we decide if the feature is allowed is the parent document's origin, and `src` is computed from the iframe's src. I was expecting it to work because the top document is allowed all permissions by default, therefore an iframe that it creates should be allowed, too. But this is wrong, in retrospect.